File: src/Aspire.Hosting.CodeGeneration.Elixir/Resources/watch.exs
# File watcher for the Aspire Elixir AppHost.
#
# This file is copied verbatim into `.aspire/modules/`. It has no Hex
# dependency and it loads none of the other SDK files.
#
#     elixir .aspire/modules/watch.exs apphost.exs
#
# The watcher starts the AppHost, forwards its output, and polls the mtime of
# every `*.ex` and `*.exs` file below the AppHost directory. A change stops the
# child with SIGTERM and starts it again.
#
# ## Options
#
#   * `--once` — do not restart. The watcher stops with the exit code of the
#     child. The tests use it to check the exit code.
#
# ## Behaviour
#
# The watcher follows `nodemon`. When the AppHost stops on its own, the watcher
# stays alive and waits for the next file change. `_build/`, `deps/` and every
# directory whose name starts with a dot, such as `.aspire/`, are not watched.

defmodule Aspire.Watch do
  @moduledoc false

  @poll_interval 500
  @stop_timeout 5_000
  @ignored_directories ["_build", "deps"]

  def main(argv) do
    {once?, arguments} = take_flag(argv, "--once")

    case arguments do
      [app_host_file | child_arguments] ->
        start(app_host_file, child_arguments, once?)

      [] ->
        IO.puts(:stderr, "usage: elixir watch.exs <apphost.exs> [--once]")
        System.halt(2)
    end
  end

  defp start(app_host_file, child_arguments, once?) do
    elixir = find_elixir()
    app_host_path = Path.expand(app_host_file)
    root = Path.dirname(app_host_path)

    trap_signals()

    state = %{
      elixir: elixir,
      app_host_file: app_host_file,
      app_host_path: app_host_path,
      root: root,
      child_arguments: child_arguments,
      once?: once?,
      port: nil,
      last_status: nil,
      snapshot: snapshot(app_host_path, root)
    }

    schedule_poll()
    loop(%{state | port: spawn_child(state)})
  end

  # ── Loop ──────────────────────────────────────────────────────────────────

  defp loop(state) do
    receive do
      {port, {:data, data}} when port == state.port ->
        IO.binwrite(data)
        loop(state)

      {port, {:exit_status, status}} when port == state.port ->
        child_stopped(state, status)

      :poll ->
        schedule_poll()
        poll(state)

      {:watch_signal, _signal} ->
        shutdown(state)

      _other ->
        loop(state)
    end
  end

  defp poll(state) do
    snapshot = snapshot(state.app_host_path, state.root)

    case first_change(state.snapshot, snapshot) do
      nil ->
        loop(%{state | snapshot: snapshot})

      changed ->
        IO.puts(:stderr, "[aspire-watch] restarting: #{relative(changed, state.root)}")
        state = %{state | snapshot: snapshot}
        state = stop_child(state)
        loop(%{state | port: spawn_child(state)})
    end
  end

  # A child that stops on its own does not stop the watcher. `nodemon` prints
  # "app crashed - waiting for file changes" and waits, and so does this.
  defp child_stopped(%{once?: true}, status), do: System.halt(status)

  defp child_stopped(state, status) do
    IO.puts(
      :stderr,
      "[aspire-watch] apphost stopped with status #{status}. Waiting for a file change."
    )

    loop(%{state | port: nil})
  end

  defp shutdown(state) do
    state = stop_child(state)
    System.halt(exit_code(state))
  end

  defp exit_code(%{last_status: status}) when is_integer(status), do: status
  defp exit_code(_state), do: 0

  # ── Child process ─────────────────────────────────────────────────────────

  defp spawn_child(state) do
    Port.open(
      {:spawn_executable, state.elixir},
      [
        :binary,
        :exit_status,
        :hide,
        :stderr_to_stdout,
        args: [state.app_host_file | state.child_arguments]
      ]
    )
  end

  defp stop_child(%{port: nil} = state), do: state

  defp stop_child(state) do
    case Port.info(state.port, :os_pid) do
      {:os_pid, os_pid} -> signal(os_pid, "TERM")
      nil -> :ok
    end

    status = await_exit(state.port, @stop_timeout)
    %{state | port: nil, last_status: status}
  end

  defp await_exit(port, timeout) do
    receive do
      {^port, {:data, data}} ->
        IO.binwrite(data)
        await_exit(port, timeout)

      {^port, {:exit_status, status}} ->
        status
    after
      timeout ->
        force_stop(port)
        nil
    end
  end

  defp force_stop(port) do
    case Port.info(port, :os_pid) do
      {:os_pid, os_pid} ->
        signal(os_pid, "KILL")

        receive do
          {^port, {:exit_status, status}} -> status
        after
          1_000 -> nil
        end

      nil ->
        nil
    end
  end

  defp signal(os_pid, name) do
    _ = System.cmd("kill", ["-#{name}", Integer.to_string(os_pid)], stderr_to_stdout: true)
    :ok
  rescue
    _exception -> :ok
  end

  # ── File snapshot ─────────────────────────────────────────────────────────

  defp snapshot(app_host_path, root) do
    [app_host_path | watched_files(root)]
    |> Enum.uniq()
    |> Map.new(fn path -> {path, mtime(path)} end)
  end

  defp watched_files(root) do
    root
    |> Path.join("**/*.{ex,exs}")
    |> Path.wildcard()
    |> Enum.reject(&ignored?(&1, root))
  end

  # Path.wildcard/1 does not match a directory whose name starts with a dot, so
  # `.aspire/modules/` never reaches this function. `_build/` and `deps/` do.
  defp ignored?(path, root) do
    path
    |> relative(root)
    |> Path.split()
    |> Enum.any?(fn segment ->
      segment in @ignored_directories or String.starts_with?(segment, ".")
    end)
  end

  defp mtime(path) do
    case File.stat(path, time: :posix) do
      {:ok, %File.Stat{mtime: mtime, size: size}} -> {mtime, size}
      {:error, _reason} -> nil
    end
  end

  defp first_change(previous, current) do
    changed =
      Enum.find(current, fn {path, stamp} -> Map.get(previous, path, :missing) != stamp end)

    case changed do
      {path, _stamp} -> path
      nil -> Enum.find_value(previous, fn {path, _} -> not_in(current, path) end)
    end
  end

  defp not_in(current, path), do: if(Map.has_key?(current, path), do: nil, else: path)

  defp relative(path, root), do: Path.relative_to(path, root)

  # ── Helpers ───────────────────────────────────────────────────────────────

  defp take_flag(argv, flag) do
    {matches, rest} = Enum.split_with(argv, &(&1 == flag))
    {matches != [], rest}
  end

  defp schedule_poll, do: Process.send_after(self(), :poll, @poll_interval)

  # The BEAM keeps SIGINT for its break handler, so only SIGTERM and SIGQUIT can
  # be trapped. Ctrl-C sends SIGINT to the full process group, so the AppHost
  # gets it directly.
  defp trap_signals do
    watcher = self()

    for signal <- [:sigterm, :sigquit] do
      System.trap_signal(signal, fn ->
        send(watcher, {:watch_signal, signal})
        :ok
      end)
    end

    :ok
  end

  defp find_elixir do
    case System.find_executable("elixir") do
      nil ->
        IO.puts(:stderr, "[aspire-watch] cannot find the elixir executable on PATH")
        System.halt(2)

      path ->
        path
    end
  end
end

Aspire.Watch.main(System.argv())

File: src/Aspire.Hosting.CodeGeneration.Dart/Resources/watch.dart
// File watcher for the Aspire Dart AppHost.
//
// This file is copied verbatim into `.aspire/modules/`. It uses the Dart SDK
// libraries only, and it loads none of the other SDK files. `aspire.dart` never
// imports it: the CLI runs it as the parent process of the AppHost when watch
// mode is on.
//
//     dart run .aspire/modules/watch.dart apphost.dart
//
// The watcher starts the AppHost with inherited stdio, then polls the
// modification time of `apphost.dart`, of `pubspec.yaml` and of every `*.dart`
// file below the AppHost directory. A change stops the child with SIGTERM, and
// with SIGKILL after five seconds, and starts it again.
//
// ## Options
//
//   * `--once` — do not restart. The watcher stops with the exit code of the
//     child. The tests use it to check the exit code.
//
// ## Behaviour
//
// The watcher follows `nodemon`. When the AppHost stops on its own, the watcher
// stays alive and waits for the next file change. `.dart_tool/`, `build/` and
// every directory whose name starts with a dot, such as `.aspire/`, are not
// watched.

import 'dart:async';
import 'dart:io';

/// The time between two snapshots of the watched files.
const Duration _pollInterval = Duration(milliseconds: 500);

/// The time the watcher waits for SIGTERM before it sends SIGKILL.
const Duration _stopTimeout = Duration(seconds: 5);

/// Directory names the watcher never descends into.
const List<String> _ignoredDirectories = <String>['.dart_tool', 'build'];

/// The manifest the watcher follows beside the AppHost file.
const String _manifestFileName = 'pubspec.yaml';

Future<void> main(List<String> arguments) async {
  final bool once = arguments.contains('--once');
  final List<String> rest = arguments
      .where((String argument) => argument != '--once')
      .toList();

  if (rest.isEmpty) {
    stderr.writeln('usage: dart run watch.dart <apphost.dart> [--once]');
    exit(2);
  }

  await _Watcher(
    appHostFile: rest.first,
    childArguments: rest.sublist(1),
    once: once,
  ).run();
}

class _Watcher {
  _Watcher({
    required this.appHostFile,
    required this.childArguments,
    required this.once,
  }) : appHostPath = File(appHostFile).absolute.path,
       root = File(appHostFile).absolute.parent.path;

  /// The AppHost path as the caller wrote it. The child gets the same text.
  final String appHostFile;

  /// The arguments that go to the AppHost after its own path.
  final List<String> childArguments;

  /// True when the watcher stops with the exit code of the first child.
  final bool once;

  /// The absolute AppHost path. The snapshot is keyed by absolute paths.
  final String appHostPath;

  /// The directory the watcher polls.
  final String root;

  Map<String, String> _snapshot = const <String, String>{};
  Process? _child;
  int? _childStatus;
  bool _reportedStop = false;

  Future<void> run() async {
    _installSignalHandlers();
    _snapshot = _takeSnapshot();
    await _startChild();

    while (true) {
      await Future<void>.delayed(_pollInterval);

      final int? status = _childStatus;
      if (status != null) {
        if (once) {
          exit(status);
        }
        if (!_reportedStop) {
          _reportedStop = true;
          stderr.writeln(
            '[aspire-watch] apphost stopped with status $status. '
            'Waiting for a file change.',
          );
        }
      }

      final Map<String, String> current = _takeSnapshot();
      final String? changed = _firstChange(_snapshot, current);
      _snapshot = current;

      if (changed == null) {
        continue;
      }

      stderr.writeln('[aspire-watch] restarting: ${_relative(changed)}');
      await _stopChild();
      await _startChild();
    }
  }

  // ── Shutdown ──────────────────────────────────────────────────────────────

  /// Stops the AppHost when the watcher itself is asked to stop.
  ///
  /// Without this the child outlives the watcher, because the CLI stops the
  /// watcher and never learns the process identifier of the AppHost.
  void _installSignalHandlers() {
    for (final ProcessSignal signal in <ProcessSignal>[
      ProcessSignal.sigterm,
      ProcessSignal.sigint,
    ]) {
      signal.watch().listen((ProcessSignal _) {
        unawaited(_shutdown());
      });
    }
  }

  Future<void> _shutdown() async {
    await _stopChild();
    exit(0);
  }

  // ── Child process ─────────────────────────────────────────────────────────

  Future<void> _startChild() async {
    _childStatus = null;
    _reportedStop = false;

    // Platform.resolvedExecutable is the `dart` binary that runs this script,
    // so the child uses the same SDK and no PATH lookup is necessary.
    final Process child = await Process.start(
      Platform.resolvedExecutable,
      <String>['run', appHostFile, ...childArguments],
      workingDirectory: root,
      mode: ProcessStartMode.inheritStdio,
    );

    _child = child;
    unawaited(
      child.exitCode.then((int code) {
        if (identical(_child, child)) {
          _childStatus = code;
        }
      }),
    );
  }

  Future<void> _stopChild() async {
    final Process? child = _child;
    _child = null;
    _childStatus = null;

    if (child == null) {
      return;
    }

    child.kill(ProcessSignal.sigterm);
    try {
      await child.exitCode.timeout(_stopTimeout);
    } on TimeoutException {
      child.kill(ProcessSignal.sigkill);
      await child.exitCode;
    }
  }

  // ── File snapshot ─────────────────────────────────────────────────────────

  Map<String, String> _takeSnapshot() {
    final Map<String, String> stamps = <String, String>{};
    _stamp(stamps, appHostPath);
    _stamp(stamps, '$root${Platform.pathSeparator}$_manifestFileName');
    _collect(Directory(root), stamps);
    return stamps;
  }

  void _collect(Directory directory, Map<String, String> stamps) {
    List<FileSystemEntity> entries;
    try {
      entries = directory.listSync(followLinks: false);
    } on FileSystemException {
      return;
    }

    for (final FileSystemEntity entry in entries) {
      final String name = _baseName(entry.path);

      if (entry is Directory) {
        if (name.startsWith('.') || _ignoredDirectories.contains(name)) {
          continue;
        }
        _collect(entry, stamps);
        continue;
      }

      if (entry is File && name.endsWith('.dart')) {
        _stamp(stamps, entry.path);
      }
    }
  }

  static void _stamp(Map<String, String> stamps, String path) {
    final File file = File(path);
    try {
      final FileStat stat = file.statSync();
      if (stat.type == FileSystemEntityType.notFound) {
        return;
      }
      stamps[file.absolute.path] =
          '${stat.modified.microsecondsSinceEpoch}:${stat.size}';
    } on FileSystemException {
      // A file that disappeared between the listing and the stat is a change
      // the next snapshot reports.
    }
  }

  /// Returns the first path that is new, changed or removed, or null.
  static String? _firstChange(
    Map<String, String> previous,
    Map<String, String> current,
  ) {
    for (final String path in current.keys.toList()..sort()) {
      if (previous[path] != current[path]) {
        return path;
      }
    }
    for (final String path in previous.keys.toList()..sort()) {
      if (!current.containsKey(path)) {
        return path;
      }
    }
    return null;
  }

  String _relative(String path) {
    final String prefix = '$root${Platform.pathSeparator}';
    return path.startsWith(prefix) ? path.substring(prefix.length) : path;
  }

  static String _baseName(String path) {
    final int index = path.lastIndexOf(Platform.pathSeparator);
    return index < 0 ? path : path.substring(index + 1);
  }
}
